Cython/Compiler/Code.py

```diff
@@ -1423,23 +1423,33 @@ def new_string_const_cname(self, bytes_value):
         value = bytes_value.decode('ASCII', 'ignore')
         return self.new_const_cname(value=value)
 
-    def new_num_const_cname(self, value, py_type):
+    def make_unique_const_cname(self, format_str):  # type: (str) -> str
+        used = self.const_cnames_used
+        cname = format_str % ""
+        counter = 0
+        while used.get(cname):
+            counter += 1
+            cname = format_str % counter
+        used[cname] = 1
+        return cname
+
+    def new_num_const_cname(self, value, py_type):  # type: (str, str) -> str
         if py_type == 'long':
             value += 'L'
             py_type = 'int'
         prefix = Naming.interned_prefixes[py_type]
-        cname = "%s%s" % (prefix, value)
-        cname = cname.replace('+', '_').replace('-', 'neg_').replace('.', '_')
+
+        value = value.replace('.', '_').replace('+', '_').replace('-', 'neg_')
+        if len(value) > 42:
+            cname = self.make_unique_const_cname(
+                prefix + "large%s_" + value[:21] + "_XXX_" + value[-21:])
+        else:
+            cname = "%s%s" % (prefix, value)
         return cname
 
     def new_const_cname(self, prefix='', value=''):
         value = replace_identifier('_', value)[:32].strip('_')
-        used = self.const_cnames_used
-        name_suffix = value
-        while name_suffix in used:
-            counter = used[value] = used[value] + 1
-            name_suffix = '%s_%d' % (value, counter)
-        used[name_suffix] = 1
+        name_suffix = self.make_unique_const_cname(value + "_%s")
```

Comment:
This appends an underscore to all these cnames. We should avoid that in the normal (already unique) case. (Note that we should remove the underscore before testing for uniqueness for the first time.)